include/deal.II/base/mpi_remote_point_evaluation.h

```diff
@@ -488,84 +702,118 @@ namespace Utilities
         Utilities::MPI::this_mpi_process(tria->get_communicator());
 
       // allocate memory for output and buffer
-      output.resize(point_ptrs.back());
-      buffer.resize(std::max(send_permutation.size() * 2,
-                             point_ptrs.back() + send_permutation.size()));
+      output.resize(point_ptrs.back() * n_components);
+
+      buffer.resize(
+        (sort_data ? buffer_size_with_sorting : buffer_size_without_sorting) *
+        n_components);
 
       // ... for evaluation
-      ArrayView<T> buffer_eval(buffer.data(), send_permutation.size());
+      ArrayView<T> buffer_eval(buffer.data(),
+                               send_permutation.size() * n_components);
 
-      // ... for communication
-      ArrayView<T> buffer_comm(buffer.data() + send_permutation.size(),
-                               send_permutation.size());
+      // ... for communication (send)
+      ArrayView<T> buffer_send(sort_data ?
+                                 (buffer.data() +
+                                  send_permutation.size() * n_components) :
+                                 (buffer.data()),
```

Comment:
I would personally remove the braces here, but this might be a matter of taste.
```suggestion
                                 buffer.data(),
```